src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Logging/VisualStudioFeedbackLoggerProvider.InMemoryBuffer.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Logging;
+
+internal partial class VisualStudioFeedbackLoggerProvider
+{
+    /// <summary>
+    /// A circular in memory buffer to store logs in memory. The intent is
+    /// to lazily allocate the buffer as needed, provide a <see cref="ToString"/> method
+    /// for human readability of logs, and prepare the memory for garbage collection
+    /// on call to <see cref="Clear"/>
+    /// </summary>
+    private class InMemoryBuffer(int bufferSize)
+    {
+        private Lazy<string[]> _memory = new (() => new string[bufferSize]);
+
+        // Start at -1 because append always increments, so we want to start at value 0
+        private int _head = -1;
+
+        public void Append(string s)
+        {
+            var position = Interlocked.Increment(ref _head) % _memory.Value.Length;
+            _memory.Value[position] = s;
+        }
+
+        public void Clear()
+        {
+            Interlocked.Exchange(ref _memory, new Lazy<string[]>(() => new string[bufferSize]));
+        }
+
+        public override string ToString()
```

Comment:
> eg, we could dump it to a text file to be uploaded as part of feedback reporting, so we don't have to download huge memory dumps to access it?

This is the part I'm super unclear on, but may be the best approach in the long run. The difference in how some files are uploaded to feedback is confusing to me. I'm going to spend some time to poke experts next week to validate my understanding. 

For example, at beginning of feedback loghub makes a zip. That zip doesn't change when you click record steps. However, that may be a specific implementation detail. We might be able to continue writing to a file instead of in memory and then have feedback grab that file on submission. 

On the `ToString` portion, honestly it helped me validate my own things. Maybe even make it easy for me to write unit tests (which I haven't done) to validate the buffer is written correctly. Agree it could not be run in a dump